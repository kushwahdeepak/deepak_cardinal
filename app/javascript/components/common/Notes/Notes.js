import React, { useState, useEffect } from 'react'
import Container from 'react-bootstrap/Container'
import Row from 'react-bootstrap/Row'
import Col from 'react-bootstrap/Col'
import Button from 'react-bootstrap/Button'
import Card from 'react-bootstrap/Card'
import Alert from 'react-bootstrap/Alert'
import AddNote from '../AddNote/AddNote'
import moment from 'moment'
import axios from 'axios'
import feather from 'feather-icons'

import styles from './styles/Notes.module.scss'

function Notes({ notes, candidate, setNotes }) {
    const [showAddNoteModal, setShowAddNoteModal] = useState(false)
    const [noteAdded, setNoteAdded] = useState(false)
    const [errorFetchingNotes, setErrorFetchingNotes] = useState(null)

    useEffect(() => {
        axios
            .get(`/people/${candidate.id}.json`)
            .then((res) => {
                setNotes(res.data.notes)
            })
            .catch((e) => setErrorFetchingNotes(e.message))
    }, [noteAdded])

    useEffect(() => {
        feather.replace()
    })

    return (
        <Container>
            {errorFetchingNotes && (
                <Alert
                    variant="danger"
                    onClose={() => setErrorFetchingNotes(null)}
                    dismissible
                >
                    {errorFetchingNotes}
                </Alert>
            )}
            <Row className="flex-column">
                <Col xs="auto">
                    <p className={styles.text}>Notes</p>
                </Col>
                <Col xs="auto">
                    <Button
                        variant="link"
                        className={styles.text + ' p-0'}
                        data-test="open-modal-btn"
                        onClick={() => setShowAddNoteModal(true)}
                    >
                        <i data-feather="plus-circle"></i> Add new note
                    </Button>
                </Col>
            </Row>
            <AddNote
                show={showAddNoteModal}
                onHide={() => setShowAddNoteModal(false)}
                candidate={candidate}
                setNoteAdded={setNoteAdded}
            />
            {notes &&
                notes.map((note) => (
                    <Card className="mb-2" key={note.id}>
                        <Card.Body>
                            <Row>
                                <Col>
                                    <Row className="mb-2">
                                        <Col>
                                            {note.user && note.user.email}
                                        </Col>
                                        <Col xs={4}>
                                            {moment(note.created_at).format(
                                                'DD MMM, YYYY  HH:mm'
                                            )}
                                        </Col>
                                    </Row>
                                    <Row>
                                        <Col>{note.body}</Col>
                                    </Row>
                                </Col>
                            </Row>
                        </Card.Body>
                    </Card>
                ))}
        </Container>
    )
}

export default Notes
