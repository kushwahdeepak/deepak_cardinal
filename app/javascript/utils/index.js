import { act } from 'react-dom/test-utils'

export const findByTestAttr = (wrapper, attr) => {
    return wrapper.find(`[data-test='${attr}']`)
}

export const timeoutResolve = (time = 0, response = {}) =>
    new Promise((resolve) => {
        setTimeout(() => resolve(response), time)
    })

export const wait = (time = 0) => {
    const timer = timeoutResolve(time)

    return act(async () => {
        await timer
    })
}
export function emailIsValid(email) {
    return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email)
}

export function generateRandom(min, max) {
    return Math.floor(min + Math.random() * (max + 1 - min))
}

export function capitalizeFirstLetter(string) {
    return string.charAt(0).toUpperCase() + string.slice(1).toLowerCase()
}

export function createMarkedList(array) {
    if (!Array.isArray(array)) {
        return []
    }
    return array.map((candidate) => ({
        id: candidate.id,
        name: candidate.id,
        fullName: `${candidate.first_name} ${candidate.last_name}`,
        emailAddress: candidate.email_address,
        isChecked: false,
    }))
}

export function stringReplace(string,toSymbol,fromSymbol){
    return string.replace(toSymbol, fromSymbol)
}

export function capitalize(string) {
    return string.replace(/\b(\w)/g, s => s.toUpperCase());
}

export function skillIsNotEmpty(value){
    return value.trim().length > 0;
}

export function firstCharacterCapital(value){
    return value?.charAt(0)?.toUpperCase() + value?.slice(1)
}
