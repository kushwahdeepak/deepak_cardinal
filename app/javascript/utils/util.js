import isNil from 'lodash.isnil'
import isEmpty from 'lodash.isempty'

class Util {
    /**
     * Define stage values
     */
    static STAGES = ['first_interview', 'second_interview', 'offer', 'reject']

    /**
     * Tests whether a single object attribute includes the at
     * least one string from the given set.
     */
    static attributePassesStringSetFilter(object, attr, stringSet) {
        if (isNil(object[attr])) return false

        // We "OR" the tag tests: the attribute needs to contain only one tag
        return stringSet
            .map((string) => string.trim().toLowerCase())
            .reduce(
                (passes, string) =>
                    passes ||
                    Util.attributeToString(object[attr])
                        .toLowerCase()
                        .includes(string),
                false
            )
    }

    /**
     * Tests whether at least one of a set of object attributes
     * includes every word from a given piece of text.
     */
    static objectPassesTextFilter(object, attrs, text) {
        text = text.trim().toLowerCase()

        if (text === '') return true

        const words = text.split(' ')

        // Join all the given attribute names to lowercase, stringified attribute values
        const allAttrsString = attrs
            .filter((attr) => !isNil(object[attr]))
            .map((attr) => Util.attributeToString(object[attr]).toLowerCase())
            .join('')

        // Is every word included in the allAttrsString?
        return words.reduce(
            (matchesAll, word) => matchesAll && allAttrsString.includes(word),
            true
        )
    }

    static attributeToString(attr) {
        if (typeof attr === 'string') {
            return attr
        } else if (Array.isArray(attr)) {
            return attr.join(',')
        } else if (typeof attr === 'object') {
            return JSON.stringify(attr)
        } else {
            return '' + attr
        }
    }

    static numberWithCommas(x) {
        return x?.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',')
    }

    /**
     * Returns number of results in format [x] - [y] of [z]
     * @param {number} totalJobsCount Total number of jobs
     * @param {number} pageCount Total number of pages
     * @param {number} activePage Currently active page
     * @param {number} pageSize How many results to display per page
     */
    static displayNumberOfResults(
        totalJobsCount,
        pageCount,
        activePage,
        pageSize,
        totalPersonCount=null,
        candidatePage=false,
        showdisplaycount=true
    ) {
        if (totalJobsCount == 0) return '0 - 0 of 0'
        const offset = pageSize * pageCount - totalJobsCount
        const lastJobIndex =
            (activePage + 1) * pageSize > totalJobsCount
                ? totalJobsCount
                : (activePage + 1) * pageSize
        const firstJobIndex =
            activePage == 0
                ? 1
                : activePage == pageCount - 1
                ? lastJobIndex - pageSize + offset + 1
                : lastJobIndex - pageSize + 1
        return `${firstJobIndex} - ${lastJobIndex} of ${Util.numberWithCommas(
            (candidatePage && showdisplaycount) ? totalPersonCount : totalJobsCount

        )}`
    }

    /**
     * Returns a string indicating the candidate's title and current
     * company, based on available information on the Person object
     * @param {object} candidate Person object from our db
     */
    static candidateCompanyString(candidate) {
        const position = candidate.company_position
        const currentCompany = candidate.current_company

        if (isEmpty(position) && isEmpty(currentCompany)) return ''

        if (!isEmpty(position) && !isEmpty(currentCompany))
            return `${position} at ${currentCompany}`

        if (!isEmpty(position)) return position

        if (!isEmpty(currentCompany)) return `at ${currentCompany}`
    }

    /**
     * Converts the given array of objects with ids to an object whose
     * keys are the ids of the array objects.
     *
     * Every object in the given array must have a unique 'id' property defined.
     *
     * @param {Array} array
     * @returns {Object}
     */
    static arrayToObject(array) {
        const obj = {}
        array.forEach((element) => (obj[element.id] = element))
        return obj
    }

    /**
     * Makes a copy of the given array and replace the value at 'index' with 'value'
     *
     * @param {Array} array
     * @param {Number} index
     * @param {*} value
     * @returns {Array}
     */
    static replaceInArrayCopy_i(array, index, value) {
        const clone = array.slice()
        clone[index] = value
        return clone
    }

    /**
     * Makes a copy of the given array and replaces the element with the same
     * .id property as obj.id with obj
     *
     * @param {Array} array
     * @param {Object} obj
     * @returns {Array}
     */
    static replaceInArrayCopy(array, obj) {
        const index = array.findIndex((element) => element.id === obj.id)
        return Util.replaceInArrayCopy_i(array, index, obj)
    }

    static deleteInArrayCopy(array, index) {
        const copy = array.slice()
        copy.splice(index, 1)

        return copy
    }

    static handleUndefined(value){
        if (value === undefined || value === 'undefined')
        return '';
        else
        return value
    }

    static handleUndefinedFullName(firstName, lastName){
      if(firstName === "undefined" || firstName === undefined)
      return lastName
      else if(lastName === "undefined" || lastName === undefined)
      return firstName
      else
      return firstName + ' ' + lastName     }

    static capitalize(value){
         return value.charAt(0).toUpperCase() + value.slice(1).toLowerCase();
    }
}

export default Util
