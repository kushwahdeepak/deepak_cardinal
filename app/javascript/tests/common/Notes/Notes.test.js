import React from 'react';
import Notes from '../../../components/common/Notes/Notes';
import renderer from 'react-test-renderer';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { findByTestAttr, wait } from '../../../utils';
import { act } from 'react-dom/test-utils';

configure({adapter: new Adapter()});

const initialShallowRender = (props = {}) => {
	return shallow(<Notes {...props}/>);
};

const initialMountRender = (props = {}) => {
	return mount(<Notes {...props}/>);
};

const initialProps = {
	candidate: []
};

describe('Notes component',  () => {
	let wrapper;
	let fullWrapper;

	beforeEach(() => {

		const props = {
			...initialProps
		};

		wrapper = initialShallowRender(props);
		fullWrapper = initialMountRender(props);
	});

	it('should render Notes component', () => {
		expect(wrapper.exists()).toBe(true);
	});

	it('should match with a snapshot', () => {
		const component = renderer.create(<Notes />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	describe('AddNote component with props', () => {
		let wrapper;
		let fullWrapper;
		const actions = async (wrapper, _actions) => {
			await act(async () => {
				await (new Promise(resolve => setTimeout(resolve, 0)));
				_actions();
				wrapper.update();
			});
		};

		beforeEach(() => {
			const props = {
				...initialProps,
			};

			wrapper = initialShallowRender(props);
			fullWrapper = initialMountRender(props);
		});

		afterEach(() => {
			jest.clearAllMocks();
		});


		it('should open Notes modal window', () => {
			const openBtn = () => findByTestAttr(fullWrapper, 'open-modal-btn').at(0);
			const modalWrapper = () => findByTestAttr(fullWrapper, 'modal-wrapper').at(0);
			openBtn().simulate('click');

			fullWrapper.update();
			expect(modalWrapper().exists()).toBeTruthy();
		});

		it('should have expected content in Notes modal window', () => {
			const expectedTitle = 'Add New Note';
			const expectedButton = 'Add';
			const openBtn = () => findByTestAttr(fullWrapper, 'open-modal-btn').at(0);
			const modalTitle = () => findByTestAttr(fullWrapper, 'notes-modal-title').at(0);
			const addModalBtn = () => findByTestAttr(fullWrapper, 'add-modal-btn').at(0);
			openBtn().simulate('click');

			const titleText = modalTitle().text();
			const addBtnText = addModalBtn().text();
			fullWrapper.update();
			expect(titleText).toBe(expectedTitle);
			expect(addBtnText).toBe(expectedButton);
		});

		it('should close Notes modal window', async () => {
			const openBtn = () => findByTestAttr(fullWrapper, 'open-modal-btn').at(0);
			const closeButton = () => findByTestAttr(fullWrapper, 'close-button').at(0);
			const modalWrapper = () => findByTestAttr(fullWrapper, 'modal-wrapper').at(0);
			openBtn().simulate('click');
			closeButton().simulate('click');

			fullWrapper.update();
			expect(modalWrapper().exists()).toBeFalsy();
		});

		it.skip('should call onClick handler', async () => {
			const setShowAddNoteModal = jest.fn();
			const handleClick = jest.spyOn(React, "useState");
			handleClick.mockImplementation(size => [size, setShowAddNoteModal]);
			const openBtn = () => findByTestAttr(wrapper, 'open-modal-btn').at(0);
			const closeButton = () => findByTestAttr(fullWrapper, 'close-button').at(0);

			openBtn().simulate('click');
			expect(setShowAddNoteModal).toHaveBeenCalledTimes(1);
		});
	});
});
