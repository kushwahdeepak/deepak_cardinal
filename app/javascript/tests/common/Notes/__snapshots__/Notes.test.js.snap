// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notes component should match with a snapshot 1`] = `
<div
  className="container"
>
  <div
    className="justify-content-between row"
  >
    <div
      className="col-auto"
    >
      Notes
    </div>
    <div
      className="col-auto"
    >
      <button
        className="p-0 btn btn-link"
        data-test="open-modal-btn"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Add Note
      </button>
    </div>
  </div>
</div>
`;
